drivers/input/input_pinnacle.c

```diff
@@ -0,0 +1,827 @@
+
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
+static bool pinnacle_is_ready_spi(const struct pinnacle_bus *bus)
+{
+	if (!spi_is_ready_dt(&bus->spi)) {
+		LOG_ERR("SPI bus %s is not ready", bus->spi.bus->name);
+		return false;
+	}
+
+	return true;
+}
+
+static int pinnacle_write_spi(const struct pinnacle_bus *bus, uint8_t address, uint8_t value)
+{
+	uint8_t tx_data[] = {
+		PINNACLE_WRITE_REG(address),
+		value,
+	};
+	const struct spi_buf tx_buf[] = {{
+		.buf = tx_data,
+		.len = ARRAY_SIZE(tx_data),
+	}};
+	const struct spi_buf_set tx_set = {
+		.buffers = tx_buf,
+		.count = ARRAY_SIZE(tx_buf),
+	};
+
+	return spi_write_dt(&bus->spi, &tx_set);
+}
+
+static int pinnacle_seq_write_spi(const struct pinnacle_bus *bus, uint8_t *address, uint8_t *value,
+				  uint8_t count)
+{
+	uint8_t tx_data[count * 2];
+	const struct spi_buf tx_buf[] = {{
+		.buf = tx_data,
+		.len = ARRAY_SIZE(tx_data),
+	}};
+	const struct spi_buf_set tx_set = {
+		.buffers = tx_buf,
+		.count = ARRAY_SIZE(tx_buf),
+	};
+
+	for (uint8_t i = 0; i < count; ++i) {
+		tx_data[i * 2] = PINNACLE_WRITE_REG(address[i]);
+		tx_data[i * 2 + 1] = value[i];
+	}
+
+	return spi_write_dt(&bus->spi, &tx_set);
+}
+
+static int pinnacle_read_spi(const struct pinnacle_bus *bus, uint8_t address, uint8_t *value)
+{
+	uint8_t tx_data[] = {
+		PINNACLE_READ_REG(address),
+		PINNACLE_SPI_FB,
+		PINNACLE_SPI_FB,
+		PINNACLE_SPI_FB,
+	};
+	const struct spi_buf tx_buf[] = {{
+		.buf = tx_data,
+		.len = ARRAY_SIZE(tx_data),
+	}};
+	const struct spi_buf_set tx_set = {
+		.buffers = tx_buf,
+		.count = ARRAY_SIZE(tx_buf),
+	};
+
+	uint8_t rx_data[4];
+	const struct spi_buf rx_buf[] = {{
+		.buf = rx_data,
+		.len = ARRAY_SIZE(rx_data),
+	}};
+	const struct spi_buf_set rx_set = {
+		.buffers = rx_buf,
+		.count = ARRAY_SIZE(rx_buf),
+	};
+
+	int rc;
+
+	rc = spi_transceive_dt(&bus->spi, &tx_set, &rx_set);
+	if (rc) {
+		LOG_ERR("Failed to read from SPI %s", bus->spi.bus->name);
+		return rc;
+	}
+
+	*value = rx_data[3];
+
+	return 0;
+}
+
+static int pinnacle_seq_read_spi(const struct pinnacle_bus *bus, uint8_t address, uint8_t *buf,
+				 uint8_t count)
+{
+
+	uint8_t size = count + 3;
+	uint8_t tx_data[size];
+
+	tx_data[0] = PINNACLE_READ_REG(address);
+	tx_data[1] = PINNACLE_SPI_FC;
+	tx_data[2] = PINNACLE_SPI_FC;
+
+	uint8_t i = 3;
+
+	for (; i < (count + 2); ++i) {
+		tx_data[i] = PINNACLE_SPI_FC;
+	}
+
+	tx_data[i++] = PINNACLE_SPI_FB;
+
+	const struct spi_buf tx_buf[] = {{
+		.buf = tx_data,
+		.len = size,
+	}};
+	const struct spi_buf_set tx_set = {
+		.buffers = tx_buf,
+		.count = 1,
+	};
+
+	uint8_t rx_data[size];
+	const struct spi_buf rx_buf[] = {{
+		.buf = rx_data,
+		.len = size,
+	}};
+	const struct spi_buf_set rx_set = {
+		.buffers = rx_buf,
+		.count = 1,
+	};
+
+	int rc;
+
+	rc = spi_transceive_dt(&bus->spi, &tx_set, &rx_set);
+	if (rc) {
+		LOG_ERR("Failed to read from SPI %s", bus->spi.bus->name);
+		return rc;
+	}
+
+	for (i = 0; i < count; ++i) {
+		buf[i] = rx_data[i + 3];
+	}
```

Comment:
I think that you can rework `rx_buf` to have two entries where the first has `.buf = NULL, len = 3` and then on the second you can use `.buf = rx_data, len = 3`, then you won't need this copy and can use the buffer directly, there's an example in `input_pmw3610.c`, can you give that a go?

Reply:
I'll give it another try because when I tried it at the beginning either the nrf driver was not supporting this or I made a mistake.

Reply:
Yes, it's working, updated the code.